### PR TITLE
GH-178: Fixes edge case of a infinite loop in decrypting properties

### LINKED ISSUE #178: EnvironmentDecryptApplicationInitializer.insert(...) infinite loop possibility
The ApplicationContext overload of insert ([currently at line 124](https://github.com/spring-cloud/spring-cloud-commons/blob/master/spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializer.java#L124)) has a while loop that looks like it could get stuck.

This probably can't happen in reality, but it's easy to fix, just move the 
```
				parent = parent.getParent();
````
one line down, **outside** the if-instance-of block:
```
		while (parent != null) {
			if (parent.getEnvironment() instanceof ConfigurableEnvironment) {
				ConfigurableEnvironment mutable = (ConfigurableEnvironment) parent
						.getEnvironment();
				insert(mutable.getPropertySources(), propertySource);
			}
			parent = parent.getParent();
		}
```

### PR DESCRIPTION
**Pull request by bijukunjummen**
Fixes #178 : Edge case where if there is a parent context which returns an environment that is not a `ConfigurableEnvironment`, can lead to an infinite loop.

### CODE REVIEW

File: spring-cloud-context/src/test/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializerTests.java [hunk 1]
```diff
@@ -32,6 +34,8 @@
 import static org.springframework.boot.test.util.EnvironmentTestUtils.addEnvironment;
 import static org.springframework.cloud.bootstrap.encrypt.EnvironmentDecryptApplicationInitializer.DECRYPTED_PROPERTY_SOURCE_NAME;
 
+import static org.mockito.Mockito.*;
```
Comment (spencergibb) [inline spring-cloud-context/src/test/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializerTests.java]: No `*` imports please.

File: spring-cloud-context/src/test/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializerTests.java [hunk 2]
```diff
@@ -115,4 +119,88 @@ public void indexedPropertiesCopied() {
...
+	public void testDecryptNonStandardParent() {
+		ConfigurableApplicationContext ctx = new AnnotationConfigApplicationContext();
+		EnvironmentDecryptApplicationInitializer initializer =
+				new EnvironmentDecryptApplicationInitializer(Encryptors.noOpText());
+
+		addEnvironment(ctx, "key:{cipher}value");
+
+
+		ApplicationContext ctxParent = mock(ApplicationContext.class);
+		when(ctxParent.getEnvironment()).thenReturn(new Environment() {
```
Comment (spencergibb) [inline spring-cloud-context/src/test/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializerTests.java]: Use `MockEnvironment` here.

File: spring-cloud-context/src/test/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializerTests.java (hunk 1 above)
Comment (bijukunjummen) [inline spring-cloud-context/src/test/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializerTests.java]: Fixed, thanks

File: spring-cloud-context/src/test/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializerTests.java (hunk 2 above)
Comment (bijukunjummen) [inline spring-cloud-context/src/test/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializerTests.java]: Not sure if that will work @spencergibb, this issue is a real corner case and can only happen if the environment is not a `ConfigurableEnvironment`. `MockEnvironment` is a `ConfigurableEnvironment` so will not simulate the bug?

Comment (spencergibb) [inline spring-cloud-context/src/test/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializerTests.java]: Good point, didn't realize that.

File: spring-cloud-context/src/test/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializerTests.java [hunk 3]
```diff
@@ -15,23 +15,27 @@
  */
 package org.springframework.cloud.bootstrap.encrypt;
 
+import static org.hamcrest.Matchers.is;
```
Comment (spencergibb) [inline spring-cloud-context/src/test/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializerTests.java]: statics to the bottom

Comment (bijukunjummen) [inline spring-cloud-context/src/test/java/org/springframework/cloud/bootstrap/encrypt/EnvironmentDecryptApplicationInitializerTests.java]: Fixed @spencergibb

**Comment by bijukunjummen** [comment]
Hi @spencergibb, I think all your review comments have now been addressed. A quick question: is this typically the right way to send in changes based on review comments, I have been committing, then squashing and pushing. Artem from the spring integration who I worked with last week has suggested not squashing on submitting updates for a PR.

**Comment by spencergibb** [comment]
@bijukunjummen doesn't matter much. We can squash easily too. You can see individual changes if you don't squash. So there is more context that way.
